addons/sale/models/product_product.py

```diff
@@ -42,6 +153,54 @@ def _onchange_type(self):
                 'message': _("You cannot change the product's type because it is already used in sales orders.")
             }}
 
+    def _can_inverse_sol_qyt(self, sale_order_line):
+        if len(sale_order_line) > 1:
+            raise UserError(
+                _("Cannot modify quantity: product present in several order lines.")
+            )
+        if sale_order_line.product_uom and sale_order_line.product_uom != self.uom_id:
+            raise UserError(
+                _("Cannot modify quantity: UoM mismatch between sale order line and product.")
+            )
+
+    def _write_sale_order_line(self, sale_order, sale_order_line):
+        qty = float_round(self.sol_qty, precision_rounding=self.uom_id.rounding)
+        sale_order_is_not_readonly = sale_order.state in ['draft', 'sent']
+
+        if qty > 0 and not sale_order_line:  # positive quantity and no line -> create one
+            sol = self.env['sale.order.line'].create({
+                'name': self.name,
+                'order_id': sale_order.id,
+                'price_unit': self._get_price_unit(sale_order),
+                'product_id': self.id,
+                'product_uom_qty': qty,
+                'product_uom': self.uom_id.id,
+                'sequence': ((sale_order.order_line and
+                                sale_order.order_line[-1].sequence + 1)
+                            or 10),  # put it at the end of the order
+            })
+            # computing name after creation,to benefit from sale_order_line logic
+            sol._compute_name()
+        elif qty != 0 and sale_order_line:
+            sale_order_line.write({'product_uom_qty': qty})
+        elif qty == 0 and sale_order_line:  # remove existing line if possible, otherwise set to zero
+            if sale_order_is_not_readonly:
+                sale_order_line.unlink()
+            else:
+                sale_order_line.write({'product_uom_qty': 0})
+
+    def get_sale_order(self):
+        order_id = self.env.context.get('order_id')
+        return self.env['sale.order'].search([('id', '=', order_id)])
+
+    def set_sol_qty(self, qty=1, click_on_record=True):
+        self.ensure_one()
+        # specify if clicked on the buttons of '+' and '-' or somewhere else on the record
+        self.sol_qty = self.sol_qty + 1 if click_on_record else qty
+
+    def remove_from_order(self):
+        self.sol_qty = 0
```

Comment:
I'd like to handle that in js if possible, it's not really great to manage that on the server.

Reply:
See https://github.com/odoo-dev/odoo/pull/2469